compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp

```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
-  auto checkEquivalent = [&](Value lhsValue, Value rhsValue) {
-    if (operands.contains(lhsValue) && operands.contains(rhsValue)) {
-      return success();
-    }
-    return success(equivalentValues.lookup(lhsValue) == rhsValue ||
-                   equivalentValues.lookup(rhsValue) == lhsValue);
+static bool checkOperationEquivalence(Operation *aOp, Operation *bOp) {
+  auto aLinalgOp = dyn_cast<linalg::LinalgOp>(aOp);
+  auto bLinalgOp = dyn_cast<linalg::LinalgOp>(bOp);
+
+  if (!aLinalgOp || !bLinalgOp) {
+    return false;
+  }
+  // Contraction ops verifies that there are two operands and one result.
+  assert(linalg::isaContractionOpInterface(aLinalgOp) &&
+         linalg::isaContractionOpInterface(bLinalgOp) &&
+         "expected lhs and rhs to be contraction ops");
+
+  // CHeck that the LHS operand is the same. Can be relaxed in the future.
+  if (aLinalgOp.getDpsInputOperand(0)->get() !=
+      bLinalgOp.getDpsInputOperand(0)->get()) {
+    return false;
+  }
+
+  auto checkSameRankAndElementType = [](Value aVal, Value bVal) {
+    auto aType = dyn_cast<ShapedType>(aVal.getType());
+    auto bType = dyn_cast<ShapedType>(bVal.getType());
+    return aType && bType && aType.getRank() == bType.getRank() &&
+           aType.getElementType() == bType.getElementType();
   };
-  auto markEquivalent = [&](Value v1, Value v2) { equivalentValues[v1] = v2; };
-  return OperationEquivalence::isEquivalentTo(
-      lhsOp, rhsOp, checkEquivalent, markEquivalent,
-      /*flags=*/OperationEquivalence::IgnoreLocations);
-}
+  // Check that the RHS rank and element type are the same. We dont check the
+  // type cause we allow RHS to be transposes.
+  if (!checkSameRankAndElementType(aLinalgOp.getDpsInputOperand(1)->get(),
+                                   bLinalgOp.getDpsInputOperand(1)->get())) {
+    return false;
+  }
 
-/// Check that an operation is a `empty -> fill -> contraction`
-static bool isEmptyFillContractionDAGRootOp(
-    linalg::LinalgOp linalgOp,
-    std::optional<linalg::LinalgOp> seedContractionOp = std::nullopt) {
-  if (!linalg::isaContractionOpInterface(linalgOp)) {
+  // Check that the output rank and element type are the same. We dont check the
+  // type cause we allow RHS to be transposes.
+  if (!checkSameRankAndElementType(aLinalgOp.getDpsInitOperand(0)->get(),
+                                   bLinalgOp.getDpsInitOperand(0)->get())) {
     return false;
   }
-  auto fillOp = linalgOp.getDpsInits()[0].getDefiningOp<linalg::FillOp>();
-  if (!fillOp) {
+
+  // Check that the iterator types are the same.
+  if (aLinalgOp.getIteratorTypesArray() != bLinalgOp.getIteratorTypesArray()) {
     return false;
   }
-  // For convenience check that the fill value is 0. This is not
-  // a necessity, but easier to handle the rewrite this way.
-  if (!matchPattern(fillOp.getDpsInputOperand(0)->get(), m_AnyZeroFloat()) &&
-      !matchPattern(fillOp.getDpsInputOperand(0)->get(), m_Zero())) {
+
+  // Check region equivalence.
```

Comment:
out of curiosity, is it a hard limit, or it can be relaxed in the future as well?

Reply:
I think it could be relaxed... but starting with this for now.